File: topk-rs/tests/test_query_vector.rs
use float8::F8E4M3;
use test_context::test_context;
use topk_rs::proto::v1::control::{FieldSpec, VectorDistanceMetric};
use topk_rs::query::{field, fns, select};

mod utils;
use topk_rs::{doc, schema};
use utils::dataset;
use utils::ProjectTestContext;

use crate::utils::is_sorted;

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_query_vector_distance(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let result = ctx
        .client
        .collection(&collection.name)
        .query(
            select([("title", field("title"))])
                .select([(
                    "summary_distance",
                    fns::vector_distance("summary_embedding", vec![2.0; 16]),
                )])
                .topk(field("summary_distance"), 3, true),
            None,
            None,
        )
        .await
        .expect("could not query");

    assert!(is_sorted(&result, "summary_distance"));
    assert_fields!(&result, ["_id", "title", "summary_distance"]);
    assert_doc_ids!(result, ["1984", "pride", "mockingbird"]);
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_query_vector_distance_without_refine(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let raw = ctx
        .client
        .collection(&collection.name)
        .query(
            select([("title", field("title"))])
                .select([(
                    "summary_distance",
                    fns::vector_distance("summary_embedding", vec![2.0; 16]).skip_refine(true),
                )])
                .topk(field("summary_distance"), 3, true),
            None,
            None,
        )
        .await
        .expect("could not query");

    assert!(is_sorted(&raw, "summary_distance"));
    assert_doc_ids!(&raw, ["1984", "pride", "mockingbird"]);

    let refined = ctx
        .client
        .collection(&collection.name)
        .query(
            select([("title", field("title"))])
                .select([(
                    "summary_distance",
                    fns::vector_distance("summary_embedding", vec![2.34; 16]),
                )])
                .topk(field("summary_distance"), 3, true),
            None,
            None,
        )
        .await
        .expect("could not query");

    assert!(is_sorted(&refined, "summary_distance"));
    assert_doc_ids!(&refined, ["1984", "pride", "mockingbird"]);

    // The refined result should be different from the raw result.
    assert_ne!(raw, refined);
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_query_vector_distance_nullable(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let result = ctx
        .client
        .collection(&collection.name)
        .query(
            select([(
                "summary_distance",
                fns::vector_distance("nullable_embedding", vec![3.0f32; 4]),
            )])
            .topk(field("summary_distance"), 3, true),
            None,
            None,
        )
        .await
        .expect("could not query");

    assert!(is_sorted(&result, "summary_distance"));
    assert_doc_ids!(result, ["1984", "mockingbird", "catcher"]);
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_query_vector_distance_u8_vector(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let result = ctx
        .client
        .collection(&collection.name)
        .query(
            select([(
                "summary_distance",
                fns::vector_distance("scalar_embedding", vec![8u8; 16]),
            )])
            .topk(field("summary_distance"), 3, true),
            None,
            None,
        )
        .await
        .expect("could not query");

    assert!(is_sorted(&result, "summary_distance"));
    assert_doc_ids!(result, ["harry", "1984", "catcher"]);
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_query_vector_distance_i8_vector(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let result = ctx
        .client
        .collection(&collection.name)
        .query(
            select([(
                "summary_distance",
                fns::vector_distance("scalar_i8_embedding", vec![-10i8; 4]),
            )])
            .topk(field("summary_distance"), 3, true),
            None,
            None,
        )
        .await
        .expect("could not query");

    assert!(is_sorted(&result, "summary_distance"));
    assert_doc_ids!(result, ["pride", "1984", "gatsby"]);
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_query_vector_distance_binary_vector(ctx: &mut ProjectTestContext) {
    let collection = dataset::books::setup(ctx).await;

    let result = ctx
        .client
        .collection(&collection.name)
        .query(
            select([(
                "summary_distance",
                fns::vector_distance("binary_embedding", vec![0u8, 1]),
            )])
            .topk(field("summary_distance"), 2, true),
            None,
            None,
        )
        .await
        .expect("could not query");

    assert!(is_sorted(&result, "summary_distance"));
    assert_doc_ids!(result, ["1984", "mockingbird"]);
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_query_vector_distance_f16_vector(ctx: &mut ProjectTestContext) {
    let collection = ctx
        .client
        .collections()
        .create(
            ctx.wrap("vec_f16"),
            schema!(
                "embedding" => FieldSpec::f16_vector(16, true, VectorDistanceMetric::Euclidean),
            ),
        )
        .await
        .expect("could not create collection");

    let lsn = ctx
        .client
        .collection(&collection.name)
        .upsert(vec![
            doc!("_id" => "1", "embedding" => vec![half::f16::from_f32(1.0); 16]),
            doc!("_id" => "2", "embedding" => vec![half::f16::from_f32(2.0); 16]),
            doc!("_id" => "3", "embedding" => vec![half::f16::from_f32(3.0); 16]),
            doc!("_id" => "4", "embedding" => vec![half::f16::from_f32(4.0); 16]),
            doc!("_id" => "5", "embedding" => vec![half::f16::from_f32(5.0); 16]),
        ])
        .await
        .expect("could not insert");

    let result = ctx
        .client
        .collection(&collection.name)
        .query(
            select([(
                "dist",
                fns::vector_distance("embedding", vec![half::f16::from_f32(3.0); 16]),
            )])
            .topk(field("dist"), 3, true),
            Some(lsn),
            None,
        )
        .await
        .expect("could not query");

    assert!(is_sorted(&result, "dist"));
    assert_doc_ids!(result, ["2", "3", "4"]);
}

#[test_context(ProjectTestContext)]
#[tokio::test]
async fn test_query_vector_distance_f8_vector(ctx: &mut ProjectTestContext) {
    let collection = ctx
        .client
        .collections()
        .create(
            ctx.wrap("vec_f16"),
            schema!(
                "embedding" => FieldSpec::f8_vector(16, true, VectorDistanceMetric::Euclidean),
            ),
        )
        .await
        .expect("could not create collection");

    let lsn = ctx
        .client
        .collection(&collection.name)
        .upsert(vec![
            doc!("_id" => "1", "embedding" => vec![F8E4M3::from_f32(1.0); 16]),
            doc!("_id" => "2", "embedding" => vec![F8E4M3::from_f32(2.0); 16]),
            doc!("_id" => "3", "embedding" => vec![F8E4M3::from_f32(3.0); 16]),
            doc!("_id" => "4", "embedding" => vec![F8E4M3::from_f32(4.0); 16]),
            doc!("_id" => "5", "embedding" => vec![F8E4M3::from_f32(5.0); 16]),
        ])
        .await
        .expect("could not insert");

    let result = ctx
        .client
        .collection(&collection.name)
        .query(
            select([(
                "dist",
                fns::vector_distance("embedding", vec![F8E4M3::from_f32(3.0); 16]),
            )])
            .topk(field("dist"), 3, true),
            Some(lsn),
            None,
        )
        .await
        .expect("could not query");

    assert!(is_sorted(&result, "dist"));
    assert_doc_ids!(result, ["2", "3", "4"]);
}

File: topk-py/tests/test_query_vector.py
import numpy as np
import pytest

from topk_sdk.data import (
    binary_vector,
    f16_vector,
    f32_vector,
    f8_vector,
    u8_vector,
    i8_vector,
)
from topk_sdk.query import field, fn, select

from . import ProjectContext
from .utils import dataset, doc_ids, is_sorted, doc_fields


def test_query_vector_distance(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    result = ctx.client.collection(collection.name).query(
        select(
            "title",
            summary_distance=fn.vector_distance("summary_embedding", [2.0] * 16),
        ).topk(field("summary_distance"), 3, True)
    )

    assert is_sorted(result, "summary_distance")
    assert doc_fields(result) == {"_id", "title", "summary_distance"}
    assert doc_ids(result) == {"1984", "pride", "mockingbird"}


def test_query_vector_distance_numpy_f32(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    result = ctx.client.collection(collection.name).query(
        select(
            "title",
            summary_distance=fn.vector_distance(
                "summary_embedding", np.array([2.0] * 16, dtype=np.float32)
            ),
        ).sort(field("summary_distance"), True).limit(3)
    )

    assert is_sorted(result, "summary_distance")
    assert doc_ids(result) == {"1984", "pride", "mockingbird"}


def test_query_vector_distance_nullable(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    result = ctx.client.collection(collection.name).query(
        select(
            summary_distance=fn.vector_distance(
                "nullable_embedding", f32_vector([3.0] * 16)
            )
        ).topk(field("summary_distance"), 3, True)
    )

    assert is_sorted(result, "summary_distance")
    assert doc_ids(result) == {"1984", "mockingbird", "catcher"}


def test_query_vector_distance_u8_vector(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    result = ctx.client.collection(collection.name).query(
        select(
            summary_distance=fn.vector_distance("scalar_embedding", u8_vector([8] * 16))
        ).topk(field("summary_distance"), 3, True)
    )

    assert is_sorted(result, "summary_distance")
    assert doc_ids(result) == {"harry", "1984", "catcher"}


def test_query_vector_distance_numpy_u8(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    result = ctx.client.collection(collection.name).query(
        select(
            summary_distance=fn.vector_distance(
                "scalar_embedding", np.array([8] * 16, dtype=np.uint8)
            )
        ).sort(field("summary_distance"), True).limit(3)
    )

    assert is_sorted(result, "summary_distance")
    assert doc_ids(result) == {"harry", "1984", "catcher"}


def test_query_vector_distance_i8_vector(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    result = ctx.client.collection(collection.name).query(
        select(
            summary_distance=fn.vector_distance(
                "scalar_i8_embedding", i8_vector([-10] * 16)
            )
        ).topk(field("summary_distance"), 3, True)
    )

    assert is_sorted(result, "summary_distance")
    assert doc_ids(result) == {"pride", "1984", "gatsby"}


def test_query_vector_distance_numpy_i8(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    result = ctx.client.collection(collection.name).query(
        select(
            summary_distance=fn.vector_distance(
                "scalar_i8_embedding", np.array([-10] * 16, dtype=np.int8)
            )
        ).sort(field("summary_distance"), True).limit(3)
    )

    assert is_sorted(result, "summary_distance")
    assert doc_ids(result) == {"pride", "1984", "gatsby"}


def test_query_vector_distance_binary_vector(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    result = ctx.client.collection(collection.name).query(
        select(
            summary_distance=fn.vector_distance(
                "binary_embedding", binary_vector([0, 1])
            )
        ).topk(field("summary_distance"), 2, True)
    )

    assert is_sorted(result, "summary_distance")
    assert doc_ids(result) == {"1984", "mockingbird"}


def test_query_vector_distance_numpy_binary(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    result = ctx.client.collection(collection.name).query(
        select(
            summary_distance=fn.vector_distance(
                "binary_embedding", np.array([0, 1], dtype=np.uint8)
            )
        ).sort(field("summary_distance"), True).limit(2)
    )

    assert is_sorted(result, "summary_distance")
    assert doc_ids(result) == {"1984", "mockingbird"}


def test_query_vector_distance_f8_vector(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    result = ctx.client.collection(collection.name).query(
        select(
            summary_distance=fn.vector_distance("f8_embedding", f8_vector([1.0] * 16))
        ).topk(field("summary_distance"), 3, True)
    )

    assert is_sorted(result, "summary_distance")
    assert doc_ids(result) == {"mockingbird", "1984", "pride"}


def test_query_vector_distance_f16_vector(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    result = ctx.client.collection(collection.name).query(
        select(
            summary_distance=fn.vector_distance("f16_embedding", f16_vector([1.0] * 16))
        ).topk(field("summary_distance"), 3, True)
    )

    assert is_sorted(result, "summary_distance")
    assert doc_ids(result) == {"mockingbird", "1984", "pride"}


def test_query_vector_distance_numpy_f16(ctx: ProjectContext):
    collection = dataset.books.setup(ctx)

    result = ctx.client.collection(collection.name).query(
        select(
            summary_distance=fn.vector_distance(
                "f16_embedding", np.array([1.0] * 16, dtype=np.float16)
            )
        ).sort(field("summary_distance"), True).limit(3)
    )

    assert is_sorted(result, "summary_distance")
    assert doc_ids(result) == {"mockingbird", "1984", "pride"}


def test_query_vector_distance_numpy_2d_invalid():
    with pytest.raises(
        ValueError, match="Vector query must be a vector or sparse vector"
    ):
        fn.vector_distance(
            "summary_embedding", np.array([[1.0, 2.0]], dtype=np.float32)
        )


def test_query_vector_distance_numpy_unsupported_dtype():
    with pytest.raises(TypeError, match="Unsupported numpy dtype"):
        fn.vector_distance("summary_embedding", np.array([1.0, 2.0], dtype=np.float64))
